s3prl/upstream/byol_a/expert.py

```diff
@@ -80,50 +73,11 @@ def forward(self, wavs):
                 each feat is in torch.FloatTensor and already
                 put in the device assigned by command-line args
         """
-        wavs_len = [len(wav) for wav in wavs]
-        max_wav_len = max(wavs_len)
-        start_points = list(range(0, max_wav_len, int(self.stride_secs * SAMPLE_RATE)))
-        padded_max_wav_len = start_points[-1] + int(self.window_secs * SAMPLE_RATE)
-        padded_wavs = [
-            torch.cat([wav, wav.new_zeros(padded_max_wav_len - len(wav))])
-            for wav in wavs
-        ]
-
-        all_features = []
-        for start in start_points:
-            subwavs = [
-                wav[start : start + int(self.window_secs * SAMPLE_RATE)]
-                for wav in padded_wavs
-            ]
-            features = [
-                self.normalizer(
-                    (self.to_melspec(wav) + torch.finfo(torch.float).eps).log()
-                ).permute(1, 0)
-                for wav in subwavs
-            ]
-            features = torch.stack(
-                features, dim=0
-            )  # (batch_size, segment_seq_len, hiddqen_size)
-            all_features.append(features)
-
-        all_features = torch.stack(all_features, dim=0)
-        num_segment, batch_size, segment_seq_len, hidden_size = all_features.shape
-
-        flatten_features = all_features.reshape(-1, segment_seq_len, hidden_size)
-
-        repre = self.model(
-            flatten_features.transpose(1, 2).unsqueeze(1)
-        )  # repre: (num_segment * batch_size, hidden_size)
-        repre = repre.reshape(num_segment, batch_size, -1).transpose(
-            0, 1
-        )  # repre: (batch_size, num_segment, hidden_size)
-
-        trimmed_hs = []
-        for h in [repre]:
-            max_h_len = len(range(0, max_wav_len, self.get_downsample_rates()))
-            h = h[:, :max_h_len, :]
-            trimmed_hs.append(h)
-
+        self.to_logmelspec.to(wavs[0].device)
+        wavs = pad_sequence(wavs, batch_first=True)
+        features = self.normalizer(self.to_logmelspec(wavs)).unsqueeze(1) # (B, F, T) -> (B, 1, F, T)
+        layered_features = self.model.by_layers(self.model(features, layered=True)) # [(B, T, D)] x 5
```

Comment:
I have a quick question just to make sure, is the `(B, T, D)` here represents `batch_size`, `number of frames` and `hidden_size`?
If so, is the `number of frames` are in order, so that the later frames always represent the later samples of the input waveform?

Reply:
Yes, your guesses are right. BYOL-A is a simple CNN that convolves spectrogram frames in order and keeps the order in the output features.

Reply:
Cool!